Refactor QAE using the Grover operator

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Part of the Finance refactor, where the `GroverOperator` is used in both Grover's algorithm and all Amplitude Estimation algorithms.

### Details and comments

Updates the signature to comply with the `GroverOperator` names and the plan for Grover's algorithm and uses the `GroverOperator` internally where possibly. The old tests are maintained in `test_amplitude_estimation_circuitfactory` until the circuit factories are deleted.

The tests should also be refactored at some point to test less standard cases (to be faster) and more corner cases (to test more precisely). But this is part of a later PR.

Draft because the 
* ~~docstrings need to be double-checked~~
* ~~the tests aren't completely ready yet~~
* ~~the Git history must be cleaned up~~

jlapeyre commented on qiskit/aqua/algorithms/amplitude_estimators/iqae.py:
```diff
@@ -60,10 +75,20 @@ def __init__(self, epsilon: float, alpha: float,
                 each iteration, can be 'chernoff' for the Chernoff intervals or 'beta' for the
                 Clopper-Pearson intervals (default)
             min_ratio: Minimal q-ratio (K_{i+1} / K_i) for FindNextK
```
Should "(K_{i+1} / K_i)"  be  :math:`(K_{i+1} / K_i)` ?

Cryoris replied:
yep, done in e1b0490 👍

woodsp-ibm commented on qiskit/aqua/algorithms/amplitude_estimators/ae.py:
```diff
@@ -122,14 +184,21 @@ def _evaluate_statevector_results(self, probabilities: Union[List[float], np.nda
                 y measurements with respective probabilities, in this order.
         """
+        if self._circuit is None:
+            raise AquaError('Before calling _evaluate_statevector_results, _circuit must be set. '
```
This error message again refers to calling private methods and private variables. Maybe it suffices to state that construct_circuit() must have be called to use.  I see this is a docstring on private method that the user should not be calling. If an end user ever saw this message it would be due to some internal programming error on the part of the algorithm right and not due to some incorrect sequence the user did? As such this is more an Internal Logic Error of the algorithm that presumably only the algo developer can fix?

Cryoris replied:
Yes, if users only call public methods this error only arises if something internal is broken. I think we have a few places where this "internal logic error" is the case. Would it make sense to note this somewhere? 
Now I'll just put that `constuct_circuit` must be called prior to this method because it sets the internal `_circuit` variable to give reasonable information to both the end user and the developer.

Cryoris replied:
Updated in 1def8c675

woodsp-ibm commented on qiskit/aqua/algorithms/amplitude_estimators/ae.py:
```diff
@@ -351,16 +419,17 @@ def loglikelihood(a):
+        # check if A factory or state_preparation has been set
+        if self.state_preparation is None:
+            if self.a_factory is None:  # getter emits deprecation warnings, therefore nest
+                raise AquaError('The A operator must be set!')
```
Why does this not say that `state_preparation must be set` or `state_preparation (A operator) must be set` if you want to include A operator. In params it calls it a circuit so switching to calling it A operator may not be so obvious to users.
In seeing the AEAlg base class below now I see that is more how state_preparation param is described there (which is a bit different to the text in this class)

Cryoris replied:
Good point, In didn't update that after updating the input arguments, thanks 👍

Cryoris replied:
Updated in fdda99467

woodsp-ibm commented on qiskit/aqua/algorithms/amplitude_estimators/iqae.py:
```diff
@@ -280,15 +349,16 @@ def _clopper_pearson_confint(self, counts: int, shots: int, alpha: float
+        # check if A factory or state_preparation has been set
+        if self.state_preparation is None:
+            if self.a_factory is None:  # getter emits deprecation warnings, therefore nest
+                raise AquaError('The A operator must be set!')
```
Similar comment to earlier

Cryoris replied:
Updated in fdda99467

woodsp-ibm commented on qiskit/aqua/algorithms/amplitude_estimators/mlae.py:
```diff
@@ -152,17 +206,25 @@ def _evaluate_statevectors(self,
         Returns:
             The corresponding probabilities.
         """
+        if self._circuits is None:
```
See earlier comment to similar code

Cryoris replied:
Updated in 1def8c675

woodsp-ibm commented on qiskit/aqua/algorithms/amplitude_estimators/mlae.py:
```diff
@@ -152,17 +206,25 @@ def _evaluate_statevectors(self,
             statevectors: A list of statevectors.
 
+        Raises:
+            AquaError: If self._circuits is None.
```
See earlier comment to similar

Cryoris replied:
Updated in 1def8c675

woodsp-ibm commented on qiskit/aqua/algorithms/amplitude_estimators/mlae.py:
```diff
@@ -407,9 +468,10 @@ def _run_mle(self) -> float:
+        # check if A factory or state_preparation has been set
+        if self.state_preparation is None:
+            if self.a_factory is None:  # getter emits deprecation warnings, therefore nest
+                raise AquaError('The A operator must be set!')
```
See earlier comment on similar code

Cryoris replied:
Updated in 1def8c675

Cryoris replied:
Updated  in fdda99467

woodsp-ibm commented on qiskit/aqua/circuits/phase_estimation_circuit.py:
```diff
@@ -129,7 +130,13 @@ def construct_circuit(
                     q = QuantumRegister(self._operator.num_qubits, name='q')
                 elif self._unitary_circuit_factory is not None:
-                    q = QuantumRegister(self._unitary_circuit_factory.num_target_qubits, name='q')
+                    if isinstance(self._unitary_circuit_factory, CircuitFactory):
```
Could we change the method signature so the types are now typehints rather than in the docstring since this is being edited. This is partly motivated since it seems that the unitary_circuit_factory may be other than a CircuitFactory type now and the latter is all that is stated as the type.

Cryoris replied:
Good point, added in 69175a230

Cryoris replied:
This functionality has been removed in 1adaed0 as it wasn't needed anymore 👍

woodsp-ibm commented on qiskit/aqua/circuits/phase_estimation_circuit.py:
```diff
@@ -141,11 +148,21 @@ def construct_circuit(
-                    num_aux_qubits = self._state_in_circuit_factory.required_ancillas()
+                    if isinstance(self._state_in_circuit_factory, CircuitFactory):
+                        num_aux_qubits = self._state_in_circuit_factory.required_ancillas()
+                    elif hasattr(self._state_in_circuit_factory, 'num_ancilla_qubits'):
+                        num_aux_qubits = self._state_in_circuit_factory.num_ancilla_qubits
```
I will note that the unit test coverage went down on PhaseEstimation with a number of the new additions, such as this, not having any unit test coverage. Can we do something to including testing of this newly added logic

Cryoris replied:
Actually I think these lines are not being tested because they are not needed. If the A and Q operator are given as circuits, the QAE algorithm uses the `PhaseEstimation` circuit from the library instead of the component. I'll remove this functionality (and type hints).

woodsp-ibm commented on test/aqua/test_amplitude_estimation_circuitfactory.py:
```suggestion
    """Test a more difficult example, motivated from Credit Risk Analysis."""
```

woodsp-ibm commented on test/aqua/test_amplitude_estimation_circuitfactory.py:
I cannot recall whether I added this test case to get test coverage of this and/or the MultivariateProblem. I recall we had no unit testing of some of these components so borrowed some notebook logic to build out some tests. If this is moved to some circuitfactory specific test that is intended to be dropped later we will need to check that we do not lose coverage of these.

Cryoris replied:
Yes, I'm also looking into the MultivariateProblem itself after this is done so we shouldn't lose coverage of it 👍

woodsp-ibm commented:
Given the various changes to the Grover and AE algos at what point is a reno release note relevant. It could be done for here to note the changes on how these are built out.

Cryoris commented on qiskit/aqua/algorithms/amplitude_estimators/ae.py:
```diff
@@ -32,66 +32,108 @@
     """
 
     def __init__(self, num_eval_qubits: int,
+                 state_preparation: Optional[Union[QuantumCircuit, CircuitFactory]] = None,
```
This is to support an old-style call with unnamed parameters, meaning
```python
num_eval_qubits = ...
a_factory = CircuitFactory(...)
ae = AmplitudeEstimation(num_eval_qubits, a_factory)
```

woodsp-ibm replied:
Ah, ok. Nice!

Cryoris commented on qiskit/aqua/algorithms/amplitude_estimators/ae.py:
```diff
@@ -122,14 +184,21 @@ def _evaluate_statevector_results(self, probabilities: Union[List[float], np.nda
                 i.e. real(statevector * statevector.conj())[0]
 
+        Raises:
+            AquaError: If `self._circuit` is not set.
```
Updated in 1def8c675

Cryoris commented:
I added a reno 👍 

Also I removed the deprecation warnings if a `CircuitFactory` is passed, since there are still several classes based on the circuit factory (mainly `MultivariateProblem` and `UnivariateProblem`) that should be deprecated, along with a suggestion on how to use circuits instead, before raising warnings in QAE. The goal for this is this release, but I'd rather do all deprecations together.
